node/modules/services.go

```diff
@@ -73,14 +79,45 @@ func RunBlockSync(h host.Host, svc *blocksync.BlockSyncService) {
 	h.SetStreamHandler(blocksync.BlockSyncProtocolID, svc.HandleStream)
 }
 
-func HandleIncomingBlocks(mctx helpers.MetricsCtx, lc fx.Lifecycle, ps *pubsub.PubSub, s *chain.Syncer, bserv dtypes.ChainBlockService, chain *store.ChainStore, stmgr *stmgr.StateManager, h host.Host, nn dtypes.NetworkName) {
-	ctx := helpers.LifecycleCtx(mctx, lc)
+func waitForSync(stmgr *stmgr.StateManager, epochs int, subscribe func()) {
+	nearsync := uint64(epochs) * uint64(build.BlockDelaySecs) * uint64(time.Second) //nolint
 
-	blocksub, err := ps.Subscribe(build.BlocksTopic(nn)) //nolint
-	if err != nil {
-		panic(err)
+	// early check, are we synced at start up?
+	ts := stmgr.ChainStore().GetHeaviestTipSet()
+	timestamp := ts.MinTimestamp()
+	now := uint64(build.Clock.Now().UnixNano())
+	if timestamp > now-nearsync {
+		subscribe()
+		return
 	}
 
+	// we are not synced, subscribe to head changes and wait for sync
+	stmgr.ChainStore().SubscribeHeadChanges(func(rev, app []*types.TipSet) error {
+		if len(app) == 0 {
+			return nil
+		}
+
+		latest := app[0].MinTimestamp()
+		for _, ts := range app[1:] {
+			timestamp := ts.MinTimestamp()
+			if timestamp > latest {
+				latest = timestamp
+			}
+		}
+
+		now := uint64(build.Clock.Now().UnixNano())
+		if latest > now-nearsync {
+			subscribe()
+			return store.ErrNotifieeDone
+		}
+
+		return nil
+	})
+}
+
+func HandleIncomingBlocks(mctx helpers.MetricsCtx, lc fx.Lifecycle, ps *pubsub.PubSub, s *chain.Syncer, bserv dtypes.ChainBlockService, chain *store.ChainStore, stmgr *stmgr.StateManager, h host.Host, nn dtypes.NetworkName) {
```

Comment:
Given that we probably want to subscribe early in bootstrapper mode, we could use the dtypes.Bootstrapper boolean here to decide if we should subscribe immediately, which should allow us drop the rather ugly global PubsubSubscribeImmediately

Reply:
Yeah, I don't like that flag one bit, but I had to fix the tests somehow :)
And yes, it makes sense to subscribe early in the bootstrappers so that we can PX to other (synced) peers, even if we get tons of Ignores.

Reply:
done, we got rid of the flag!